### ISSUE TITLE
Question about Multi-class focal loss

### ISSUE THREAD

**Issue by DingYuan0118** [CoinCheung/pytorch-loss#29]
I notice that it seems that [`focal_loss.py`](https://github.com/CoinCheung/pytorch-loss/blob/master/pytorch_loss/focal_loss.py) only implements focal loss on the binary classification setting. Do there exist any implementations about focal loss for a multi-class classification setting?

**Comment by CoinCheung** [comment]
Hi,

I can think of two methods for multi-class classification:
1) focal loss works in same way like sigmoid-cross-entropy, thus you can consider each class as a seperate binary classification task. 
2) change sigmoid function into softmax function, this would require modifying the code.

From my personal experience, focal loss does not work so well as expected in multi-class classification tasks. I have not done sufficiently experiments to verify this, but I simply have this impression.

**Comment by CoinCheung** [comment]
question about answered.

**Comment by liyiersan** [comment]
> I notice that it seems that [`focal_loss.py`](https://github.com/CoinCheung/pytorch-loss/blob/master/pytorch_loss/focal_loss.py) only implements focal loss on the binary classification setting. Do there exist any implementations about focal loss for a multi-class classification setting?

I have implemented a pytorch version of Softmax Focal Loss following source code from keras.
class CategoricalFocalLoss(nn.Module):
    # keras implementation: https://github.com/keras-team/keras/blob/master/keras/src/losses/losses.py#L1587C1-L1674C21
    def __init__(self, alpha=0.25, gamma=2.0, reduction='mean'):
        super(CategoricalFocalLoss, self).__init__()
        self.alpha = alpha
        self.gamma = gamma
        self.reduction = reduction

    def forward(self, y_true, y_pred):
        """
        Computes the categorical focal crossentropy loss.

        Args:
            y_true: Tensor of true targets. shape=(batch_size) or shape=(batch_size, num_classes)
            y_pred: Tensor of predicted logits. shape=(batch_size, num_classes)

        Returns:
            Categorical focal crossentropy loss value.
        """
        
        # Ensure y_true is one-hot encoded
        if y_true.dim() == 1:
            y_true = F.one_hot(y_true, num_classes=y_pred.size(1)).float()
            
        # Apply softmax to convert logits to probabilities
        y_pred_probs = F.softmax(y_pred, dim=-1)

        # Calculate cross entropy
        cross_entropy = -y_true * torch.log(y_pred_probs)

        # Calculate modulating factor
        modulating_factor = torch.pow(1.0 - y_pred_probs, self.gamma)

        # Apply alpha weighting
        alpha_factor = y_true * self.alpha + (1 - y_true) * (1 - self.alpha)
        focal_loss = alpha_factor * modulating_factor * cross_entropy
        
        focal_loss = focal_loss.sum(dim=-1)

        if self.reduction == 'mean':
            return focal_loss.mean()
        elif self.reduction == 'sum':
            return focal_loss.sum()
        elif self.reduction == 'none':
            return focal_loss
        else:
            raise ValueError("Invalid reduction option. Choose from 'mean', 'sum', or 'none'.")

**Comment by liyiersan** [comment]
> I notice that it seems that [`focal_loss.py`](https://github.com/CoinCheung/pytorch-loss/blob/master/pytorch_loss/focal_loss.py) only implements focal loss on the binary classification setting. Do there exist any implementations about focal loss for a multi-class classification setting?

I have implemented a Pytorch verison of FocalSoftmaxLoss based source codes from keras.
```python
class CategoricalFocalLoss(nn.Module):
    # keras implementation: https://github.com/keras-team/keras/blob/master/keras/src/losses/losses.py#L1587C1-L1674C21
    def __init__(self, alpha=0.25, gamma=2.0, reduction='mean'):
        super(CategoricalFocalLoss, self).__init__()
        self.alpha = alpha
        self.gamma = gamma
        self.reduction = reduction

    def forward(self, y_true, y_pred):
        """
        Computes the categorical focal crossentropy loss.

        Args:
            y_true: Tensor of true targets. shape=(batch_size) or shape=(batch_size, num_classes)
            y_pred: Tensor of predicted logits. shape=(batch_size, num_classes)

        Returns:
            Categorical focal crossentropy loss value.
        """
        
        # Ensure y_true is one-hot encoded
        if y_true.dim() == 1:
            y_true = F.one_hot(y_true, num_classes=y_pred.size(1)).float()
            
        # Apply softmax to convert logits to probabilities
        y_pred_probs = F.softmax(y_pred, dim=-1)

        # Calculate cross entropy
        cross_entropy = -y_true * torch.log(y_pred_probs)

        # Calculate modulating factor
        modulating_factor = torch.pow(1.0 - y_pred_probs, self.gamma)

        # Apply alpha weighting
        alpha_factor = y_true * self.alpha + (1 - y_true) * (1 - self.alpha)
        focal_loss = alpha_factor * modulating_factor * cross_entropy
        
        focal_loss = focal_loss.sum(dim=-1)

        if self.reduction == 'mean':
            return focal_loss.mean()
        elif self.reduction == 'sum':
            return focal_loss.sum()
        elif self.reduction == 'none':
            return focal_loss
        else:
            raise ValueError("Invalid reduction option. Choose from 'mean', 'sum', or 'none'.")

```

**Comment by cemamxiaoxixi** [comment]
您的邮件我已经收到，会尽快回复的。
